Tech ydiago

sangallidavide opened the pull request:
Reducing duplication in ydiago implementation

sangallidavide commented:
Still ongoing. Just to let you know that I'm working on this. I'll open the PR when ready

andreamarini commented:
Dear Davide. This comment applies to both this and isolve BSE.

The BSE is a normal pseudo-hermitian matrix and I do believe that any tool that aims at diagonalizing, storing, manipulating the matrix should go in a library made available to the code.

Of course operations specific to the structure of the kernel have to be in the libraries dealing with the kernel. 

But any operation that acts on the Hamiltonian as pseudo-hermitian matrix  should go in a agnostic library.

muralidhar-nalabothula commented:
> Still ongoing. Just to let you know that I'm working on this. I'll open the PR when ready

Thanks, Davide! It’s much better now! I haven’t gone through all the changes yet, but here’s a suggestion (I’ll review it in more detail in the coming days):  

1. Move `bse/K_fill_block_cyclic.F` to `src/matrices`, possibly with a better name.  
2. Move `bse/YDIAGO_driver.F` to `src/linear_algebra`.  

**Just a comment:**  
The `YDIAGO` always gives the overlap matrix identity across all solvers (Hermitian, pseudo-Hermitian, and general non-Hermitian matrix solvers). This is achieved by setting the left eigenvectors equal to `Transpose(Inverse(Right_evs))`, thereby avoiding the need to store a very large matrix. However, note that the left eigenvectors are not normalized, while the right eigenvectors are always normalized to 1.  

In the upcoming months, I’ll work on improving the library to:  
- Introduce a better interface.  
- Remove the need for MPI compilation for both CPU and GPU (add CUDA, HIP or MAGMA solvers).  

This should eliminate a lot of code from the `linear_algebra` folder and should also partially address Andrea's concerns, if I understood them correctly.

sangallidavide commented:
Thanks for the suggestions.

For a start I'm just trying to

a) remove duplication (in the first commits I merged the "residuals" subroutines, still working on that)

b) adapt the ydiago implementation to the yambo logic, e.g. renamed the main subroutine "YDIAGO_driver" to mimic the logic of "LINEAR_ALGEBRA_driver" and "MATRIX_slepc"

As soon as I'm done, I'll let you know.

sangallidavide commented:
@muralidhar-nalabothula please check. There is still work to be done, but I've done quite some cleaning and I think it is ready for a first merge. 

Now with elpa it is possible to select either `BSSFirstEig,BSSNEig` or `BSSEnTarget,BSSNEig`
I used the variables already available for slepc

I'm running the test-suite on my machine. Wait for results here:
https://media.yambo-code.eu/robots/GPL/tech-ydiago/unimi-XPS-8930.php
If everything is ok (at present tech-ydiago has ~20 errros, let's see this new branch how it goes) and you are fine with the changes just push "merge pull request".

muralidhar-nalabothula commented on src/bse/K_diago_driver.F:
```diff
@@ -619,47 +548,79 @@ subroutine K_slepc_resize_variables()
+       !
+       if (l_target_energy) then
+         neigs_range  = 0
+         eigvals_range= (/BSS_target_E,BSS_target_E+0.01/HA2EV*BSS_n_eig/)
```
This is bit misleading. IMO the user must give two values like for example broadening in bse. I see that slepc only needs the centre and finds neig closest values to it. This is different, here we will get all eigenvalues between ```(eigvals_range(1), eigvals_range(2))```. This is bit confusing for the user as he might except that the only gets 100 eigenvalues but might endup with fewer or more.

muralidhar-nalabothula commented on src/bse/YDIAGO_driver.F:
```diff
@@ -236,7 +243,9 @@ subroutine K_diagonalize(i_BS_mat, BS_energies, BS_VR, &
       !
-      if(present(neigs_range))  neig_found = abs(maxval(neigs_range))
+      if(present(neigs_range)) then 
+         if (neigs_range(1) > 0) neig_found = abs(maxval(neigs_range))
```
Warn the user if ```neigs_range(1) > 1 ```` . Elpa, in hermitian case can  get only eigs from [1, n] where n <= bse_dim .
In non-hermitian case, ELPA will give full spectrum irrespective to what is provided in ```neigs_range``` and ```eigvals_range```. Nevertheless the library will automatically deal with this.

Please note that ```eigvals_range``` is only for scalapack (works for both hermitian and pseudo-hermitian```. Geev solver will ignore this

muralidhar-nalabothula reviewed (approved):
Hi Davide,

Fine for me if it passes all the test-cases. I left some comments and questions. They are just for some reference, If you like address those (I would have edited them but, cannot due to permission), please go ahead or I can later fix them and open a PR.

I cannot also merge, so please go head with it

sangallidavide commented:
I addressed some of the comments from @muralidhar-nalabothula 

About the SP vs c_float, and similar, I think you are right, we should fix this. The following thread seems to be relevant:
https://stackoverflow.com/questions/55298784/any-platform-where-fortran-double-precision-is-different-from-c-double

sangallidavide commented:
Branch merged.

@muralidhar-nalabothula  , if you want to proceed with further changes, please go ahead.
